renku/core/management/session/docker.py

```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018-2021 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Docker based interactive session provider."""
+
+from pathlib import Path
+from typing import List, Optional, Tuple
+from uuid import uuid4
+
+import docker
+from yaspin import yaspin
+
+from renku.core import errors
+from renku.core.models.session import ISessionProvider
+from renku.core.plugins import hookimpl
+from renku.core.utils import communication
+
+
+class DockerSessionProvider(ISessionProvider):
```

Comment:
This class could use more typing information. Return types of methods, that `client` is `LocalClient`, `commit_sha` is `str` etc. Other classes in this PR as well.